kcidb/test_oo.py

```diff
@@ -0,0 +1,224 @@
+"""kcidb.oo test module"""
+
+import kcidb
+
+
+class KCIDBOOClientTestCase(kcidb.unittest.TestCase):
+    """Test case for oo.Client"""
+    # Conform to unittest conventions, pylint: disable=invalid-name
+    one = 1
+    two = 2
+
+    def _check_type_instance(self, item, type_name, no_instance):
+        """
+        Check the instance of a type name and that the correct
+        number of instance is returned for each type name.
+        Args:
+            item:           The object returned from the oo data client.
+            type_name:      The schema type name.
+            no_instance:    Number of instance of object
+                            returned for each schema type
+        """
+        instance = item[type_name]
+        self.assertEqual(len(instance), no_instance)
+        for i in instance:
+            self.assertIsInstance(i, kcidb.oo.CLASSES[type_name])
+
+    def assertIsTestInstance(self, item, no_instance):
+        """Check that the instance of a oo.Test is returned"""
+        self._check_type_instance(item, "test", no_instance)
+
+    def assertIsBuildInstance(self, item, no_instance):
+        """Check that the instance of a oo.Build is returned"""
+        self._check_type_instance(item, "build", no_instance)
+
+    def assertIsCheckoutInstance(self, item, no_instance):
+        """Check that the instance of a oo.Checkout is returned"""
+        self._check_type_instance(item, "checkout", no_instance)
+
+    def assertIsRevisionInstance(self, item, no_instance):
+        """Check that the instance of a oo.Revision is returned"""
+        self._check_type_instance(item, "revision", no_instance)
+
+    def assertContainsAllSchemaType(self, obj, no_instance):
+        """Check that the returned object contains all the schema type."""
+        assert set(obj) == set(kcidb.oo.CLASSES)
+        for type_name in obj:
+            self._check_type_instance(obj, type_name, no_instance)
```

Comment:
These are pretty good, but they still don't check that the returned dataset doesn't contain instances of types it shouldn't contain. E.g. if we're requesting revisions, then we shouldn't get tests in the response dataset, and these functions don't test that.

I think you could replace all these with just one method checking for a set of specified types and specified number of instances, while also checking no other types are present. I also notice that you have a query for each invocation, so we can combine this into something like this:
```python
def assertResponseContains(self, pattern_str, types, num_instances):
    """
    Check that the query for the specified pattern string returns objects of types
    with specified names, in specified quantity, and nothing else.

    Args:
        ...
    """
    # For Abdulwasiu: These are not a part of the test, but are just validating the arguments
    assert isinstance(pattern_str, str)
    assert isinstance(types, set)
    assert types < set(kcidb.oo.CLASSES)
    assert isinstance(num_instances, int)

    data = self.client.query(kcidb.orm.Pattern.parse(pattern_string))

    # For Abdulwasiu: These are actual test checks
    self.assertEqual(set(data), types)
    for type_name, type_objs in data.items():
        self.assertEqual(len(type_objs), num_instances,
                         msg=f"Invalid number of {type_name} instances")
        for i, obj in enumerate(type_objs):
            self.assertIsInstance(obj, kcidb.oo.CLASSES[type_name],
                                  msg=f"Invalid type of {type_name} #{i}")
```

I renamed `no_instance` to `num_instances` for clarity, because the former could be read as "no instance", meaning "there is no instance".

Reply:
> I renamed `no_instance` to `num_instances` for clarity, because the former could be read as "no instance", meaning "there is no instance".

👍👍